Toolkit/ArcGISToolkit/TimeSlider.swift

```diff
@@ -0,0 +1,2539 @@
+// limitations under the License.
+
+import UIKit
+import QuartzCore
+import ArcGIS
+
+// MARK: - Time Slider Control
+
+/*
+ Controls how labels on the slider are displayed.
+ */
+public enum TimeSliderLabelMode {
+    case none
+    case thumbs
+    case ticks
+}
+
+/*
+ Playback direction when time is being animated.
+ */
+public enum PlaybackDirection {
+    case forward
+    case backward
+}
+
+/*
+ The looping behavior of the slider when time is being animated.
+ */
+public enum LoopMode {
+    case none
+    case `repeat`
+    case reverse
+}
+
+/*
+ Different color themes of the slider.
+ */
+public enum Theme {
+    case black
+    case blue
+    case oceanBlue
+}
+
+public class TimeSlider: UIControl {
+    
+    // MARK: - Public Properties
+    
+    // MARK: Current Extent Properties
+    
+    /*
+     The current time window or time instant of the slider
+     */
+    public var currentExtent: AGSTimeExtent? {
+        didSet {
+            if let startTime = currentExtent?.startTime, let endTime = currentExtent?.endTime {
+                //
+                // Update current extent start and end time
+                updateCurrentExtentStartTime(startTime)
+                updateCurrentExtentEndTime(endTime)
+                
+                // Update flag only if it's not an internal update
+                if !internalUpdate {
+                    //
+                    // Range is enabled only if both times are not same.
+                    isRangeEnabled = (startTime != endTime)
+                }
+            }
+            // This means there is only one thumb needs to be displayed and current extent start and end times are same.
+            else if let startTime = currentExtent?.startTime, currentExtent?.endTime == nil {
+                //
+                // Only one thumb should be displayed
+                // so set range to false
+                isRangeEnabled = false
+                
+                // Update current extent start time
+                updateCurrentExtentStartTime(startTime)
+                
+                // Start and end time must be same.
+                currentExtentEndTime = currentExtentStartTime
+            }
+            // This means there is only one thumb needs to be displayed and current extent start and end times are same.
+            else if let endTime = currentExtent?.endTime, currentExtent?.startTime == nil {
+                //
+                // Only one thumb should be displayed
+                // so set range to false
+                isRangeEnabled = false
+                
+                // Update current extent start time
+                updateCurrentExtentStartTime(endTime)
+                
+                // Start and end time must be same.
+                currentExtentEndTime = currentExtentStartTime
+            }
+            // Set start and end time to nil if
+            // current extent is nil
+            else if currentExtent == nil {
```

Comment:
@njarecha Is it possible for `currentExtent` to be non-`nil`, but for `startTime` and `endTime` to be `nil`? If so, it looks like that case might not be handled.

Reply:
fixed